File: src/test/java/io/veridia/jsonlogic/NumericComparisonExpressionTests.java
package io.veridia.jsonlogic;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class NumericComparisonExpressionTests {
  private static final JsonLogic jsonLogic = new JsonLogic();

  @Test
  public void testLessThan() {
    String json = "{\"<\" : [1, 2]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testLessThanOrEqual() {
    String json = "{\"<=\" : [1, 1]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testGreaterThan() {
    String json = "{\">\" : [2, 1]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testGreaterThanOrEqual() {
    String json = "{\">=\" : [1, 1]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testBetweenExclusive() {
    String json = "{\"<\" : [1, 2, 3]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testBetweenInclusive() {
    String json = "{\"<=\" : [1, 1, 3]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testGtBetweenExclusive() {
    String json = "{\">\" : [3, 2, 1]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testGtBetweenInclusive() {
    String json = "{\">=\" : [3, 1, 1]}";
    Object result = jsonLogic.apply(json, null);

    assertEquals(true, result);
  }

  @Test
  public void testEdgeCases() {
    assertEquals(true, jsonLogic.apply("{\">=\" : [3, 1, 1, 1]}", null));
    assertEquals(false, jsonLogic.apply("{\">=\" : [3, 1, 3, 1]}", null));
  }
}

File: src/test/scala/io/veridia/jsonlogic/NumericComparisonExpressionTests.scala
package io.veridia.jsonlogic

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class NumericComparisonExpressionTests extends AnyFunSuite with Matchers:
  private val jsonLogic = new JsonLogic()

  private val cases: Seq[(String, String, Boolean)] = Seq(
    ("less than", "{\"<\" : [1, 2]}", true),
    ("less than or equal", "{\"<=\" : [1, 1]}", true),
    ("greater than", "{\">\" : [2, 1]}", true),
    ("greater than or equal", "{\">=\" : [1, 1]}", true),
    ("chained < exclusive", "{\"<\" : [1, 2, 3]}", true),
    ("chained <= inclusive", "{\"<=\" : [1, 1, 3]}", true),
    ("chained > exclusive", "{\">\" : [3, 2, 1]}", true),
    ("chained >= inclusive", "{\">=\" : [3, 1, 1]}", true),
    ("chained >= 4 args all satisfied", "{\">=\" : [3, 1, 1, 1]}", true),
    ("chained >= 4 args one pair fails", "{\">=\" : [3, 1, 3, 1]}", false)
  )

  for (name, json, expected) <- cases do
    test(name) {
      jsonLogic.apply(json, null) shouldEqual expected
    }
